core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMNode.kt

```diff
@@ -32,18 +38,42 @@ data class STDCMNode(
 ) : Comparable<STDCMNode> {
 
     /**
-     * Defines the estimated better path between 2 nodes, in terms of total run time, then departure
-     * time, then number of reached targets. If the result is negative, the current node has a
-     * better path, and should be explored first. This method allows us to order the nodes in a
-     * priority queue, from the best path to the worst path. We then explore them in that order.
+     * Defines the estimated better path between 2 nodes, in the following priority
+     * - lowest total run time
+     * - earliest departure time
+     * - closest planned arrival time, using the current then the last planned timing data
+     * - highest number of reached targets
+     *
+     * If the result is negative, the current node has a better path, and should be explored first.
+     * This method allows us to order the nodes in a priority queue, from the best path to the worst
+     * path. We then explore them in that order.
      */
     override fun compareTo(other: STDCMNode): Int {
         val runTimeEstimation = timeSinceDeparture + remainingTimeEstimation
         val otherRunTimeEstimation = other.timeSinceDeparture + other.remainingTimeEstimation
         // Firstly, minimize the total run time: highest priority node takes the least time to
         // complete the path
-        return if (abs(runTimeEstimation - otherRunTimeEstimation) >= 1e-3)
+        return if (areTimesDifferent(runTimeEstimation, otherRunTimeEstimation))
             runTimeEstimation.compareTo(otherRunTimeEstimation)
+        // Minimise the difference with the planned arrival times
+        else if (
+            plannedTimingData != null &&
+                other.plannedTimingData != null &&
+                areTimesDifferent(
+                    plannedTimingData.arrivalTime.seconds - time,
+                    other.plannedTimingData.arrivalTime.seconds - other.time
+                )
+        )
+            (plannedTimingData.arrivalTime.seconds - time).compareTo(
+                other.plannedTimingData.arrivalTime.seconds - other.time
+            )
+        // If not, minimise the difference with the planned arrival times at the last planned node
+        else if (
+            prevPlannedNodeTimeDiff != null &&
+                other.prevPlannedNodeTimeDiff != null &&
+                areTimesDifferent(prevPlannedNodeTimeDiff, other.prevPlannedNodeTimeDiff)
+        )
+            prevPlannedNodeTimeDiff.compareTo(other.prevPlannedNodeTimeDiff)
```

Comment:
This block could be reorganized with extra variables and possibly extra methods. Complex `if` like that are difficult to read (especially with the reformating).